src/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -1,32 +1,165 @@
 import           Database.LSMTree.Internal.Run.BloomFilter (Bloom)
 import qualified Database.LSMTree.Internal.Run.BloomFilter as Bloom
 import           Database.LSMTree.Internal.Run.Index.Compact (CompactIndex,
-                     PageSpan)
+                     PageSpan (..))
 import qualified Database.LSMTree.Internal.Run.Index.Compact as Index
 import           Database.LSMTree.Internal.Serialise
+import           System.FS.API (Handle)
+import           System.FS.BlockIO.API
+
+-- | View of a 'Run' with only the essentials for lookups.
+data RunLookupView h = RunLookupView {
+    rlvKOpsHandle :: !h
+  , rlvBloom      :: !(Bloom SerialisedKey)
+  , rlvIndex      :: !CompactIndex
+  }
+  deriving Functor
 
--- | TODO: placeholder type for a run, replace by actual type once implemented
-type Run fd = (fd, Bloom SerialisedKey, CompactIndex)
+-- | Prepare disk lookups by doing bloom filter queries, index searches and
+-- creating 'IOOp's. The result is a vector of 'IOOp's and a vector of indexes,
+-- both of which are the same length. The indexes record the run and key
+-- associated with each 'IOOp'.
+prepLookups ::
+     PrimMonad m
+  => V.Vector (RunLookupView (Handle h))
+  -> V.Vector SerialisedKey
+  -> m (VU.Vector (RunIx, KeyIx), V.Vector (IOOp m h))
+prepLookups rs ks = do
+  let rkixs = bloomQueries rs ks
+  ioops <- indexSearches rs ks rkixs
+  pure (rkixs, ioops)
 
--- | Prepare disk lookups by doing bloom filter queries and index searches.
+type KeyIx = Int
+type RunIx = Int
+type ResIx = Int -- Result index
+
+-- | Perform a batch of bloom queries. The result is a tuple of indexes into the
+-- vector of runs and vector of keys respectively.
+--
+-- Note: the result vector should be ephemeral, so do not retain it.
 --
--- Note: results are grouped by key instead of file descriptor, because this
--- means that results for a single key are close together.
-prepLookups :: [Run fd] -> [SerialisedKey] -> [(SerialisedKey, (fd, PageSpan))]
-prepLookups runs ks =
-    [ (k, (fd, pspan))
-    | k <- ks
-    , r@(fd,_,_) <- runs
-    , pspan <- toList (prepLookup r k)
-    ]
-
-prepLookup :: Run fd -> SerialisedKey -> Maybe PageSpan
-prepLookup (_fd, b, fpix) k
-  | Bloom.elem k b = Index.toPageSpan $ Index.search k fpix
-  | otherwise      = Nothing
+-- TODO: we consider it likely that we could implement a optimised, batched
+-- version of bloom filter queries, which would largely replace this function.
+bloomQueries :: V.Vector (RunLookupView h) -> V.Vector SerialisedKey -> VU.Vector (RunIx, KeyIx)
+bloomQueries rs ks = VU.create $ do
+      -- The result vector can be of variable length, so we use a generous
+      -- estimate here, and we grow the vector if needed. TODO: tune the
+      -- starting estimate based on the expected true- and false-positives.
+      res <- VUM.unsafeNew (2 * ksN)
+      resixVar <- newPrimVar 0
+      res' <- go res resixVar
+      resix <- readPrimVar resixVar
```

Comment:
Why a mutable var as an argument rather than just having `resix` returned as an extra result?

Reply:
For convience, but since `bloomQueries` has been overhauled, it's disappeared